drivers/hmis/app/graphql/mutations/submit_assessment.rb

```diff
@@ -2,62 +2,69 @@ module Mutations
+      definition = assessment.assessment_detail.definition
+      enrollment = assessment.enrollment
 
-        form_definition = Hmis::Form::Definition.find_by(id: form_definition_id)
-        errors << InputValidationError.new('Form definition must exist') unless form_definition.present?
-      else
-        errors << InputValidationError.new('Enrollment ID or Assessment ID must exist', attribute: 'enrollment_id')
-      end
-
-      return { assessment: nil, errors: errors } if errors.present?
-
-      # Create new Assessment (and AssessmentDetail) if one doesn't exist already
-      assessment ||= Hmis::Hud::Assessment.new_with_defaults(
-        enrollment: enrollment,
-        user: hmis_user,
-        form_definition: form_definition,
-        assessment_date: assessment_date ? Date.strptime(assessment_date) : Date.today,
+      # Determine the Assessment Date (same as Information Date) and validate it
+      assessment_date, errors = definition.find_and_validate_assessment_date(
+        hud_values: input.hud_values,
+        entry_date: enrollment.entry_date,
+        exit_date: enrollment.exit_date,
       )
 
+      # Validate form values based on FormDefinition
+      validation_errors = definition.validate_form_values(input.hud_values, nil)
+      # If user has already confirmed any warnings, remove them
+      validation_errors = validation_errors.reject(&:warning?) if input.confirmed
+      errors.push(*validation_errors)
+
       # Update values
-      assessment.assessment_detail.assign_attributes(values: values, hud_values: hud_values)
+      assessment.assessment_detail.assign_attributes(
+        values: input.values,
+        hud_values: definition.key_by_field_name(input.hud_values),
+      )
       assessment.assign_attributes(
         user_id: hmis_user.user_id,
         date_updated: DateTime.current,
-        assessment_date: assessment_date ? Date.strptime(assessment_date) : assessment.assessment_date,
+        assessment_date: assessment_date || assessment.assessment_date,
       )
 
-      # TODO: return validation errors for processed records
+      # If this is an existing assessment and all the errors are warnings, save changes before returning.
+      # (NOTE: We could/should do this for new assessments, too, but it's a bit more complicated
+      # because we'd need to send back the newly created assessment ID to the frontend.)
+      if errors.all?(&:warning?) && assessment.id.present?
+        assessment.assessment_detail.save!
+        assessment.save!
+      end
+
+      return { assessment: nil, errors: errors } if errors.any?
+
+      # Run processor to create/update related records
+      assessment.assessment_detail.assessment_processor.run!
+
+      # Run both validations
+      assessment_valid = assessment.valid?
+      assessment_detail_valid = assessment.assessment_detail.valid?
```

Comment:
Does it make sense to _always_ validate assessment details when you validate assessment?  If so, I wonder if we should either add a validates associated in assessment?
https://guides.rubyonrails.org/v3.2/active_record_validations_callbacks.html#validates_associated

Reply:
I don't know..this would make things more complicated because AssessmentDetail also has a validates_associated specified. We do them together right now, but we might theoretically want to first check if the Assessment is valid, (maybe return early if not), and then validate all of the related records. I prefer to keep them separated right now